src/basic/terminal-util.c

```diff
@@ -1206,38 +1204,50 @@ bool terminal_is_dumb(void) {
         return getenv_terminal_is_dumb();
 }
 
-bool colors_enabled(void) {
+int colors_mode(void) {
 
-        /* Returns true if colors are considered supported on our stdout. For that we check $SYSTEMD_COLORS first
-         * (which is the explicit way to turn colors on/off). If that didn't work we turn colors off unless we are on a
-         * TTY. And if we are on a TTY we turn it off if $TERM is set to "dumb". There's one special tweak though: if
-         * we are PID 1 then we do not check whether we are connected to a TTY, because we don't keep /dev/console open
-         * continuously due to fear of SAK, and hence things are a bit weird. */
+        /* Returns the mode used to choose output colors. The value can be 0 for no colors, 16 for only the base 16
+         * ANSI colors or 256 for more colors. For that we check $SYSTEMD_COLORS first (which is the explicit way to
+         * change the mode). If that didn't work we turn colors off unless we are on a TTY. And if we are on a TTY
+         * we turn it off if $TERM is set to "dumb". There's one special tweak though: if we are PID 1 then we do not
+         * check whether we are connected to a TTY, because we don't keep /dev/console open continuously due to fear
+         * of SAK, and hence things are a bit weird. */
 
-        if (cached_colors_enabled < 0) {
-                int val;
+        if (cached_colors_mode < 0) {
+                int err, val;
 
-                val = getenv_bool("SYSTEMD_COLORS");
-                if (val >= 0)
-                        cached_colors_enabled = val;
+                err = getenv_int("SYSTEMD_COLORS", &val);
```

Comment:
I don't think this should be done with an integer parser. First of all, it breaks compat, since getenv_bool() also supports parsing "true" and "yes" and so on, which getenv_int() doesn't. But more importantly: you are not actually using this as integer, but as a simple enum, as the only values you support are 0, 1, 16, 256. 

As such, parse it as a string, and wrap parse_boolean(). i.e.:

```c
typedef enum color_mode {
        COLOR_OFF,
        COLOR_ON,
        COLOR_16,
        COLOR_256,
        _COLOR_MAX,
        _COLOR_INVALID = -1,
} color_mode;

color_mode get_color_mode(void) {
        const char *e;
        e = getenv("SYSTEMD_COLORS");
        if (!e) {
                …
        } else if (streq(e, "16")) 
                …
       else if (streq(e, "256"))
                …
        else {
                r = parse_boolean(e);
                …
        }
…
```

You get the idea

Reply:
Good idea! This further simplifies the parsing and makes the `switch`s clearer.f